crates/bin/prove_block/src/state_utils.rs

```diff
@@ -26,81 +40,205 @@ use crate::rpc_utils::CachedRpcStorage;
 
     let mut ffc_for_contract_class = get_ffc_for_contract_class_facts(&ffc);
 
-    let updated_contract_classes = match shared_state.contract_classes {
-        Some(tree) => {
-            log::debug!(
-                "Updating contract class tree with {} modifications...",
-                class_hash_to_compiled_class_hash.len()
-            );
-            let modifications: Vec<_> = class_hash_to_compiled_class_hash
-                .into_iter()
-                .map(|(key, value)| (key.to_biguint(), ContractClassLeaf::create(value)))
-                .collect();
-            Some(tree.update(&mut ffc_for_contract_class, modifications, &mut facts).await?)
+    // Update class trie with declared classes
+    let updated_global_class_root: Option<PatriciaTree> = match shared_state.contract_classes {
+        Some(class_trie) => {
+            let classes_modifications = update_empty_class_trie_with_block_incoming_changes(processed_state_update);
+            Some(class_trie.update(&mut ffc_for_contract_class, classes_modifications, &mut facts).await?)
         }
         None => {
-            assert_eq!(class_hash_to_compiled_class_hash.len(), 0, "contract_classes must be concrete before update.");
+            assert_eq!(
+                processed_state_update.class_hash_to_compiled_class_hash.len(),
+                0,
+                "contract_classes must be concrete before update."
+            );
             None
         }
     };
 
-    let accessed_addresses: HashSet<_> = accessed_addresses.into_iter().map(Felt252::from).collect();
-
     Ok((
-        accessed_addresses,
+        // TODO: we dont need to pass it as its already in processed_state_update
+        processed_state_update.accessed_addresses.clone(),
         SharedState {
             contract_states: updated_global_contract_root,
-            contract_classes: updated_contract_classes,
+            contract_classes: updated_global_class_root,
             ffc,
             ffc_for_class_hash: ffc_for_contract_class,
             contract_addresses: Default::default(),
         },
     ))
 }
+
+fn update_empty_class_trie_with_block_incoming_changes(
+    processed_state_update: &ProcessedStateUpdate,
+) -> Vec<(BigUint, ContractClassLeaf)> {
+    let classes_modifications = processed_state_update
+        .class_hash_to_compiled_class_hash
+        .iter()
+        .map(|(key, value)| (key.to_biguint(), ContractClassLeaf::create(*value)))
+        .collect();
+    classes_modifications
+}
+
+async fn update_empty_contract_state_with_block_incoming_changes(
+    empty_contract_states: &mut HashMap<TreeIndex, ContractState>,
+    processed_state_update: &ProcessedStateUpdate,
+    provider: &JsonRpcClient<HttpTransport>,
+    block_id: BlockId,
+    ffc: &mut FactFetchingContext<CachedRpcStorage, PedersenHash>,
+) -> Result<HashMap<TreeIndex, ContractState>, Box<dyn Error>> {
```

Comment:
Avoid returning `Box<dyn Error>`, it's not the way to go. Two solutions here:

1. Use `thiserror`, but it's pretty verbose for a binary application
2. Use `anyhow`.

I think we can use `anyhow` and get decent exception logging here.